Cleanup of third-party login and auto-enrollment

bradenmacdonald opened the pull request:
**Description**
This is a fix for [CRI-9](https://openedx.atlassian.net/browse/CRI-9) ("Third party auth association fails to link on login/registration") and [ECOM-1503](https://openedx.atlassian.net/browse/ECOM-1503). In order to fix that issue in a clean way, I did some refactoring of parts of the login and auto-enrollment flow.

There were three different parts of the code handling user login and optional auto-enrollment: the old login/register views, the third party auth pipeline, and the new combined "logistration" page. Each of these three had a completely separate implementation for auto-enrolling users into a course following login/registration. Also, none of them correctly completed the third party auth pipeline under all circumstances, meaning that users who do a "social" login often find that their social account never gets linked to their edX account. 

[Current Situation](https://docs.google.com/drawings/d/1OYIoXPlHst4kPav4_NxqI0xufEVv-bb3LXpzc7ShEeU/edit) vs. [New Situation](https://docs.google.com/drawings/d/1c4naj3FJ-75tjsvdK2ZjygSlV0GgfLhXoZKliYRHxII/edit)

**Detailed Description**
First, it's important to understand the many different ways that users can login to or register for the LMS:

* Can be a potential user (not yet registered), a newly registered user (email is unverified, won't be able to login again until verifying), or a normal user
* Can be authenticated via first party auth (username/password), third_party_auth (Facebook/Google/etc.), Legacy Shibboleth (work is underway to consolidate this with third_party_auth), or the edX API (used by the mobile app)
* Can use the new combined login/registration ("logistration") page (à la edx.org) or the old/deprecated login/registration pages (à la Edge). Each of which uses different APIs to do the login/registration.
* The login/register page[s] will sometimes be passed a "course_id" (and other) parameters from the edX marketing site. If present, the code must try to enroll the user in the specified course once the user has logged in or created an account. If the course is paid, the user should be redirected to the checkout flow.

The code has two significant issues that are addressed by this PR:
* Third party authentication is not handled correctly in many cases
  - As you can see in the [flow chart](https://docs.google.com/drawings/d/1OYIoXPlHst4kPav4_NxqI0xufEVv-bb3LXpzc7ShEeU/edit), there are several ways to login/register (vertical columns) where the third_party_auth ("TPA") pipeline does not get completed. This goes against [the design principles of the pipeline](https://github.com/edx/edx-platform/blob/da57b13b/common/djangoapps/third_party_auth/pipeline.py#L468) and means that user's social accounts do not get linked to their edX account during the login or registration process.
* There are three redundant implementations of auto-enrollment (see scribbled red circles on the [flow chart](https://docs.google.com/drawings/d/1OYIoXPlHst4kPav4_NxqI0xufEVv-bb3LXpzc7ShEeU/edit)):
  - third_party_auth.pipeline.change_enrollment *sometimes* handles auto-enrollment if third party auth is being used for login. It updates the CourseEnrollment table directly.
  - student.views.[create_account/login_user] handle auto-enrollment for the old login/register screens (using `try_change_enrollment` and `change_enrollment` in that same file)
  - the new login form has client-side code (AccessView.js) containing the auto-enrollment logic, and it uses the `commerce` API which in turn uses the `enrollment` API

Changes in this PR:
* python-social-auth version is bumped to 0.2.7 which gives us more fine-grained control over the pipeline when unverified users are involved. This change is confined to its own commit.
* The three different flows for auto-enrollment and all the redundant logic is combined into a new page. The login views and the third party auth views no longer need to know anything about auto-enrollment; they simply know to send the user to the "auto-enrollment" page following login/registration. The new page has the same auto-enrollment logic that was previously used by the combined logistration form. I just moved it from `AccessView.js` to `FinishAuthView.js`.
  - See [New Design flow chart](https://docs.google.com/drawings/d/1c4naj3FJ-75tjsvdK2ZjygSlV0GgfLhXoZKliYRHxII/edit)
* Adds a new "Dummy" third party auth provider that can be used for tests and development.

Benefits of this PR:
* Social auth accounts now get linked correctly during login/registration. There is no need for users to go to their "Account Settings" page and manually link their accounts.
* The social auth pipeline now always gets completed, as [required by its design assumptions](https://github.com/edx/edx-platform/blob/da57b13b/common/djangoapps/third_party_auth/pipeline.py#L468).
* No API changes were necessary
* Redundant auto-enrollment logic has been reduced substantially
* The auto-enrollment process now displays status updates to the user, which is nicer than the login form simply taking an unusually long time to submit.
* If desired, the marketing site could be changed to simply enroll users in a course by sending them to `/account/finish_auth?course_id=x&enrollment_action=enroll` regardless of whether they are logged in or not. Currently the marketing site tracks the user's logged in status and uses different APIs and different logic depending on whether or not the user is logged in. If it did this, there is even more code that could be removed from edX platform, since the login/register pages would no longer need to know about any query string arguments other than `next`.

**Sandbox**
* http://sandbox4.opencraft.com/login - updated June 18

**Upgrade Issue**
When merged, this PR does cause a problem for existing installations: `python-social-auth` at one point converted from `syncdb` to south/django migrations, and bumping the version of `python-social-auth` that we use will result in this error next time `paver update_db` is run:
```
django.db.utils.DatabaseError: table "social_auth_usersocialauth" already exists
```
To fix this (on devstack), the user must manually run this command:
```
./manage.py lms migrate default 0001_initial --fake --settings=devstack
```
Unfortunately, I'm not aware of any way to avoid this issue other than this manual fix. New installs are not affected.

**Screenshots**

* [New auto-enrollment page](https://cloud.githubusercontent.com/assets/945577/7879628/f738e534-05a6-11e5-9b44-fe56cc6fa773.png). (Only ever seen if auto-enrollment is happening, and is not really meant to be seen at all. If the server's responding quickly, the user will barely see it and will be immediately redirected.) 


**Testing Instructions**

1. Checkout this branch, get it running, and configure ~/lms.auth.json so that a social auth provider (e.g. "Google" or "Dummy") is enabled. Dummy requires no configuration and is the easiest, but it always returns the same user information/id.
2. Note the setting of `ENABLE_COMBINED_LOGIN_REGISTRATION` in the `FEATURES` of `~/lms.env.json` (this change should work whether it's true or false, but each uses different code paths)
3. Open an incognito window for the remainder of this test.
4. On your LMS, go /login . Click on a third party auth provider.
5. Once back at the login page ("You authenticated but this account is not associated with an edX account"), click "Register Now" and finish creating the account.
6. Go to Account Settings and verify that the third party account is linked. (This indicates the fix worked).
7. In your devstack console output, copy the link from the activation email, and paste it into the browser in order to activate this account. (Or run these commands in the django shell: `from django.contrib.auth.models import User; u=User.objects.get(username="Galactica1"); u.is_active=True; u.save()`)
8. From Account Settings, unlink the account and also create a password for the user.
9. Go back to /login and use the third party auth to login again. Because we unlinked, the user won't be recognized ("You authenticated but this account is not associated with an edX account"). Enter the password from step 8 and login.
10. Go to Account Settings and verify that the third party account is linked again because of this new login. (This indicates the fix worked).
11. Paste the URL /account/finish_auth?course_id=course-v1%3Aorg%2Bcourse%2Brun&enrollment_action=enroll&email_opt_in=false but replace `org` `course` and `run` with the values for an actual course on this instance.
12. Verify that the user gets auto-enrolled in that course.
13. Close and re-open the incognito window.
14. Paste the URL /login?course_id=course-v1%3Aorg%2Bcourse%2Brun&enrollment_action=enroll&email_opt_in=false but replace `org` `course` and `run` with the values for a different course on this instance.
15. Login using third party auth, and verify that the user gets auto-enrolled in that new course.
16. (In a normal browser window - not incognito:) As an admin-capable user, go to /admin/auth/user/ and delete the user you have been testing with. It's important that you delete it from that page by checking the checkbox next to that user then selecting "Delete" from the dropdown menu. If you are not on a debug machine, this may not work - in that case, use the shell to delete the UserPreference entries for the user, then delete the user using any means.
17. (Back in incognito:) Go to /login?course_id=course-v1%3Aorg%2Bcourse%2Brun&enrollment_action=enroll&email_opt_in=false but replace `org` `course` and `run` with the values for a course on this instance.
18. Click "Register". The URL may change, but that's ok.
19. Click on a third party provider, finish creating the account, then make sure that the user has been enrolled in the course and that the third party account has been linked.
20. Repeat step 16
21. Change the setting of `ENABLE_COMBINED_LOGIN_REGISTRATION`, restart the server, and repeat the process to test the other views.

**TODO**:
* A JIRA task should be created to remove/simplify `student.views.change_enrollment` in the future (the LMS should use the enrollment API instead for consistency with the login process)

**Reviewers**:
Code: Destination team
Product: Griff, Beth
DevOps: (TBD)

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=8262&repo=edx%2Fedx-platform

smarnach commented:
@bradenmacdonald I had a short look at the code, and did a bit of testing.  I'll do a full pass through the code tomorrow.

Unfortunately, the fix doesn't seem to work for me.  I went to the sandbox, clicked on "Sign in", then on "Dummy", then on "Create an account", entered some data and clicked "Create your account", went to the account settings and scrolled to the bottom.  The "Connected Accounts" list shows Dummy with the link "Link" behind it, indicating that the account is not linked yet.  Finishing the registration by clicking the activation link in the email doesn't seem to change anything.

I also got a few 500s when trying to remove the accounts I added again, but they are likely caused by me creating the superuser the wrong way (using `./manage.py lms createsuperuser --settings=aws`), so they don't seem to be related to this PR.

bradenmacdonald commented:
@smarnach  Hmm, I'll look into it. Regarding the 500 errors when deleting users: The behaviour on our sandbox seems different than on devstack. It turns out that you have to delete all the associated UserPreference objects, and then you can delete the user without issue.

bradenmacdonald commented:
@smarnach Ah, turns out a last-minute change I'd made had broken this if you started with the login page rather than the register page I believe I have fixed it now by rewriting the `ensure_user_information` pipeline step.

smarnach commented:
@bradenmacdonald Thanks, the linking works fine now.  After the steps described in my previous comment, "Dummy" is listed with an "Unlink" link after it in the user settings, indicating that linking has worked.  However, if I then immediately log out, click "Sign in" and "Dummy" again, nothing happens (well, the page gets reloaded, but that's hardly noticable).  After account activation, logging in with the Dummy button works fine.

smarnach commented:
With `ENABLE_COMBINED_LOGIN_REGISTRATION` set to `False`, I get a 500 on loading the login page.  I commented on the offending line.

smarnach commented:
I updated the code on the sandbox and tested the latest fix – works fine.  The issue that logging in with Dummy again before activating the account does nothing without giving an error message to the user still remains (but might be considered a different bug outside the scope of this fix).

Let me know when this is ready for a final round of testing.

bradenmacdonald commented:
@smarnach This is ready for you to review again. All the tests should be passing now.

Regarding the lack of a visible error message for inactive accounts: I have a fix for that on my other PR, #8155 (3c62a6be8555fe662e52019ef10c928b89a0e28f). Tomorrow I can try cherry-picking that into this PR.

bradenmacdonald commented:
@smarnach Ok, error messages should now be working too. You can test them locally if you disable django debug toolbar.

smarnach commented:
@bradenmacdonald I tested it again with various settings and login flows, and couldn't find any issues with the latest version.  :+1:

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-631](https://openedx.atlassian.net/browse/OSPR-631) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

sarina commented:
@cpennington : Would this PR be one you should review (related to SSO)? Should the Destination team also take a pass at it? Thanks for advice.

cpennington commented:
I don't think this is one that I need to review explicitly. This appears to mainly be a cleanup of the auth pipeline.

sarina commented:
@cpennington so should the destination team grab this? Or a different team?

cpennington commented:
Yeah, destination edx, or platform. It's just not strongly tied to the SSO work.

sarina commented:
OK I'll ask the Destination team to review this, since Platform already has 4-5 reviews in backlog.

sarina commented:
@bradenmacdonald your write up indicates you need ops and accessibility review - can you be explicit about what component(s) you need these teams to review?

bradenmacdonald commented:
@sarina Ops needs to be aware of the part mentioned in "Upgrade Issue": this PR bumps python-social-auth which changed from syncdb to migrations, so a manual fix must be run on the server when this code is pushed.

I think someone who works on the mobile app should review/test as well to make sure the changes to "auth_exchange" won't break the mobile login/auth process.

I removed accessibility just now - I don't think it does have any implications for them.

sarina commented:
@feanil there's a heads up for Ops regarding this PR; @nasthagiri there is some question about the mobile auth/login process as relates to this PR. Please see the above comment.

I've asked the Destination team to do the code review for this PR but your input will also be necessary. Please let me know if you'd like me to make separate tickets for your team regarding this PR.

bradenmacdonald commented:
Jenkins, test this please.

bradenmacdonald commented:
Jenkins, test this please.

rlucioni commented:
@sarina or @bradenmacdonald, has anyone on the UX team seen the [new auto-enrollment page](https://cloud.githubusercontent.com/assets/945577/7879628/f738e534-05a6-11e5-9b44-fe56cc6fa773.png)? If not, I think they'd appreciate the heads up. It's not visible for long, but I do see it when enrolling in courses on the sandbox.

rlucioni commented:
@bradenmacdonald I played around with this on your sandbox, and noticed that the account linking doesn't seem to work in the following (relatively common) scenario. Given an activated, signed-out edX account not linked to any third-party auth providers:

1) Navigate to `/login`.
2) Click "Sign in with Dummy."
3) Upon redirection to `/login`, sign in using first-party auth.
4) Once signed in, navigate to Account Settings. Under Connected Accounts, Dummy appears to remain unlinked.

Are you aware of this issue? My impression is that the changes in this PR are meant to resolve this.

In addition, assuming the problem described above is fixed, copy displayed when attempting to sign in with an unlinked third-party account will need to be updated to reflect that account linking no longer needs to be done manually. The following copy is displayed on the old login page:

![](http://i.imgur.com/qzgrQcH.png)

On the new "logistration" page, the following copy appears:

![](https://i.imgur.com/MLnM8mm.png)

I should also note that I accidentally put your sandbox into a less-than-optimal state by creating an account with the details populated by default when attempting to register with Dummy. I don't have access to that default email address, so the account remains inactive.

rlucioni commented on lms/templates/student_account/login.underscore:
```diff
@@ -1,7 +1,7 @@
     <% if (context.currentProvider) { %>
         <p class="message-copy">
-            <%- _.sprintf( gettext("You've successfully signed into %(currentProvider)s, but your %(currentProvider)s account isn't linked with an %(platformName)s account. To link your accounts, go to your %(platformName)s account settings."), context ) %>
+            <%- _.sprintf( gettext("You've successfully signed into %(currentProvider)s, but your %(currentProvider)s account isn't linked with an %(platformName)s account. To link your accounts, login now using your %(platformName)s password."), context ) %>
```
Looks like you caught the logistration page copy update.

srpearce replied:
You have successfully signed into %(currentProvider)s, but your %(currentProvider)s account is not linked with this %(platformName)s account. To link your accounts, sign in now using your %(platformName)s password.

bradenmacdonald replied:
@srpearce Would it work to say this? "You have successfully signed into %(currentProvider)s, but your %(currentProvider)s account is not linked with **any** %(platformName)s account. To link your accounts, sign in now using your %(platformName)s password."  I don't think we should say "*this* edX account" before the user has signed in to an edX account.

rlucioni commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -192,6 +176,12 @@ class AuthEntryError(AuthException):
 
 
+class NotActivatedException(AuthException):
+    """ Raised when a user tries to login to an uverified account """
```
s/uverified/unverified

bradenmacdonald replied:
Fixed. f3c43a1495b28c0ae8d9f383aa8858ed17da1e55

rlucioni commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -470,63 +445,56 @@ def ensure_user_information(strategy, auth_entry, user=None, *args, **kwargs):
+            # of the 'social' link entry occurs in one of the following pipeline steps).
+            # Reject this login attempt and tell the user to validate their account first.
+
+            # TODO: resend validation email
```
The existing SSO implementation resends this activation email. I'd consider excluding that from this work a regression.

bradenmacdonald replied:
Fixed. f3c43a1495b28c0ae8d9f383aa8858ed17da1e55

rlucioni commented on common/djangoapps/student/helpers.py:
```diff
@@ -199,3 +201,70 @@ def check_verify_status_by_course(user, course_enrollment_pairs, all_course_mode
+        try:
+            redirect_to = reverse('dashboard')
+        except NoReverseMatch:
+            redirect_to = reverse('home')
```
I don't understand why this is necessary. `lms/urls.py` doesn't include a URL named `'home'`. When would reversing the name `'dashboard'` fail?

bradenmacdonald replied:
This method is sometimes called within the CMS environment, during a run of the unit tests for the 'student' app, since student is shared by LMS and CMS. I was getting failing tests without that. Once we completely remove the legacy login/register views, and just have the ones in student_account, then this can be moved to student_account and there will be no need for it.

rlucioni replied:
Thanks for the context.

rlucioni commented on common/djangoapps/student/helpers.py:
```diff
@@ -199,3 +201,70 @@ def check_verify_status_by_course(user, course_enrollment_pairs, all_course_mode
+            redirect_to = reverse('dashboard')
+        except NoReverseMatch:
+            redirect_to = reverse('home')
+    if 'course_id' in request.GET:
```
This is inconsistent with the function's docstring, which states that "If 'course_id' is set, or other POST_AUTH_PARAMS, we will need to send the user to the /account/finish_auth/ view..." I'm also not comfortable with the assumption being made here. Namely, that the decision to perform post-auth actions is made solely on the basis of the `course_id` parameter. It may be true now, but that might not always be the case.

Ideally, we'd perform post-auth actions if any of the parameters in `POST_AUTH_PARAMS` is present in `request.GET`.

bradenmacdonald replied:
Ok, I'll make that change. I agree, other post-auth actions could be added in the future.

bradenmacdonald replied:
Fixed. f3c43a1495b28c0ae8d9f383aa8858ed17da1e55

rlucioni commented on common/djangoapps/student/tests/tests.py:
```diff
@@ -819,38 +819,6 @@ def test_change_to_honor_if_verified_not_active(self):
-        self.user = User.objects.create(username="jack", email="jack@fake.edx.org")
-
-    @unittest.skipUnless(settings.FEATURES.get('ENABLE_SHOPPING_CART'), "Shopping Cart not enabled in settings")
-    def test_change_enrollment_add_to_cart(self):
```
Why was this test removed? It doesn't appear to be related to third-party auth.

rlucioni replied:
Never mind, I answered my own question. It's because you've removed `change_enrollment()`.

rlucioni replied:
Actually, I think you should replace this test. There are two `change_enrollment` methods, one in `student/views.py` and another in `third_party_auth/pipeline.py`. You've removed the `change_enrollment` method from `third_party_auth/pipeline.py`, but this test relies on the method from `student/views.py`.

rlucioni replied:
Ugh, and now I see that you've removed the conditional in `student/views.py`'s `change_enrollment()` which is responsible for passing the request to the shoppingcart. Ignore me.

bradenmacdonald replied:
Haha, no worries, this code, and my changes to it, are pretty complex. I'm hoping that by moving all the auto-enrollment and related logic to one place, this PR will help to simplify things.

rlucioni replied:
I agree, thanks for taking this on!

feanil commented:
@bradenmacdonald So it looks like we'll have to fake the migration as a part of this upgrade.  We do red/black deploys for edx.org  will running the fake migration cause any issues to the running version of the app?  I wouldn't think it would but I want to confirm that.

@rlucioni are you the edx person who is corrdinating this PR? You'll have to make sure this info doesn't get lost and that the release master and ops person for the release after this is merged is made aware.

@sarina this should probably also be noted in any upgrade notes from birch to cyprus.

@e0d what do you think of this issue?  We could add the fake migration to the ansible scripts and then remove it in the next open source release, or we could do it manually and add it to the manual steps of just the next open source release.

sarina commented:
>  this should probably also be noted in any upgrade notes from birch to
cyprus.

Can you explain what "this" is? I can make sure the info gets to Docs
(presuming this PR merges before Cypress is cut), but I need to know what
to say - and I haven't been following this one too closely. Thanks!

On Tue, Jun 9, 2015 at 3:16 PM, Feanil Patel <notifications@github.com>
wrote:

> @bradenmacdonald <https://github.com/bradenmacdonald> So it looks like
> we'll have to fake the migration as a part of this upgrade. We do red/black
> deploys for edx.org will running the fake migration cause any issues to
> the running version of the app? I wouldn't think it would but I want to
> confirm that.
>
> @rlucioni <https://github.com/rlucioni> are you the edx person who is
> corrdinating this PR? You'll have to make sure this info doesn't get lost
> and that the release master and ops person for the release after this is
> merged is made aware.
>
> @sarina <https://github.com/sarina> this should probably also be noted in
> any upgrade notes from birch to cyprus.
>
> @e0d <https://github.com/e0d> what do you think of this issue? We could
> add the fake migration to the ansible scripts and then remove it in the
> next open source release, or we could do it manually and add it to the
> manual steps of just the next open source release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8262#issuecomment-110469468>.
>

bradenmacdonald commented:
@rlucioni That account linking issue you noticed only affects the old login page. I didn't notice the sandbox was still configured to the old login page. With the new combined login page, it works properly:

1. Account 'adama@fleet.colonies.gov' has no linked third party accounts, is activated, and has password 'edx'
2. I go to http://sandbox4.opencraft.com/login
3. I click "Dummy"
4. I see "You've successfully signed into Dummy, but your Dummy account isn't linked with an Your Platform Name Here account. To link your accounts, login now..."
5. I login with 'adama@fleet.colonies.gov' / 'edx'
6. I go to account settings, and I can see Dummy is now linked to my account.

I have changed the sandbox to the combined login page and reset the adama user to step 1 above, so you can test this if you wish. I imagine that it will be ok not to change the old login page over, since there's no behaviour regression and it's deprecated anyways, but let me know if that needs to be done. Other flows like registering with social auth are working completely on both old and new.

bradenmacdonald commented:
@feanil The tricky thing about faking the migration is that it needs to be faked if the table already exists (from when this package used syncdb), but if this is a fresh deployment then it must not be faked, or the table will never get created. Once way to achieve that would be:

```bash
./manage.py lms migrate default --settings=aws || ./manage.py lms migrate default 0001 --fake --settings=aws
```

If this is a new install, only the first command will run. If it's an existing install, the first command will error and the second will fix it. If it's an instance where third party auth is disabled, both commands will likely result in harmless errors, as the 'default' app is not even activated.

To answer your other question: I have tested what happens if you fake the migration on this branch and then revert back to `master` where the python-social-auth version is using syncdb, and there are no problems. However, you can't fake the migration before upgrading to this branch because the migration file doesn't exist. (Unless you faked it manually with `INSERT INTO south_migrationhistory (app_name, migration, applied) VALUES ("default", "0001_initial", NOW());` but you'd have to be careful about doing that only in the right circumstances).

I wish this wasn't so messy, but this is an external piece of software and I haven't yet been able to think of any nicer way to handle this upgrade.

rlucioni commented:
@bradenmacdonald got it, thank you. I've independently verified that the account linking during login works as expected. Since the issue I observed with the old login page isn't a regression, I think leaving it alone is reasonable. Overall, this looks solid. Thanks for addressing my comments so quickly.

My only remaining request is that you squash your commits before this is merged. Since the python-social-auth version bump is required by your changes, I'd include it in your squashed commit; @feanil, would you prefer that the requirements change be kept separate?

@sarina, this PR includes changes to some user-facing copy; could we arrange for Docs to take a quick look? I'd be happy to sit down with whoever represents them to help identify the affected lines. I also think the new auth completion page needs an accessibility review; I'm not sure if flashing a page for a short amount of time is an accessibility concern.

@bradenmacdonald and @sarina, this still needs eyes from someone in mobile (@nasthagiri) and someone in UX (@frrrances), as well as resolution on deployment strategy from DevOps (@feanil).

@feanil I'm representing engineering on this PR. I'll make sure to notify the appropriate parties during the release. I believe I'll also be the one verifying these changes on stage. Are you on board with the python-social-auth version bump?

bradenmacdonald commented:
@rlucioni Thanks a lot for the quick review of this big PR! I'd like to add one or two new bok choy tests later today that will confirm the third party accounts are linking during the login - I think that this was not possible before but with the new "Dummy" provider I've added we can now test these flows in bok choy. Once I've added those tests and addressed any further review issues I will definitely be squashing this down before merge. I was thinking of keeping two commits: one for the python-social-auth upgrade (a4fc1ea8a50507418839f3e167b81af01f3b4015), and one for all the rest of the changes.

rlucioni commented:
@bradenmacdonald getting those bok-choy tests in would be great. If you ping me once those are in, I'll take a last look (and bring my thumb with me).

sarina commented:
@feanil @nasthagiri please let me know if you would like an official OSPR ticket to represent your review of this pull request work. UX work is captured by [OSPR-636](https://openedx.atlassian.net/browse/OSPR-636). Docs work is captured by [DOC-2063](https://openedx.atlassian.net/browse/DOC-2063)

bradenmacdonald commented:
Jenkins, test this please

srpearce commented:
@rlucioni I'll be reviewing any doc strings in this PR. Can you point me to them?

rlucioni commented on lms/djangoapps/student_account/views.py:
```diff
@@ -170,72 +163,43 @@ def _third_party_auth_context(request):
-
-    # Check if the user is trying to enroll in a course
-    # that they don't have access to based on country
-    # access rules.
```
@bradenmacdonald I'm taking another look and noticed that the new auth completion page doesn't appear to enforce country access/embargo rules by preventing enrollment and redirecting the user to the "blocked" page. The auth completion page should verify that the authenticated user is allowed to access the requested course before enrolling them.

bradenmacdonald replied:
@rlucioni That's weird, it should be using exactly the same logic as the combined login/register page currently does. I'll look into this later today.

rlucioni replied:
@bradenmacdonald looks like we should be in the clear here. The enrollment API is ultimately responsible for enrolling the user, and it includes [enforcement of country access rules](https://github.com/edx/edx-platform/blob/master/common/djangoapps/enrollment/views.py#L412-L423).

bradenmacdonald replied:
@rlucioni That was what I had thought as well when I looked at that code. However, if you look at it again you'll see in the comment right above the part that you linked to, it says: "We do this at the view level (rather than the Python API level) because this check requires information about the HTTP request". Unfortunately, it turns out that the "commerce.baskets" API which is used by the combined login page (and now by the `finish_auth` view) uses the enrollment Python API directly, rather than the view. So the check isn't happening.

I just did some detailed tests and found that this bug exists on the existing combined login page (i.e. before this PR), and that this PR has not fixed or (nor made it worse). Here's what happens:

1. User uses a link containing ?course_id to logs i (either with the combined login page, or combined page + finish_auth)
2. the JavaScript does `POST /commerce/baskets/` to try to enroll the user in the course using the baskets API.
3. The baskets view sees that there are no SKUs for the course, so no ecommerce flow will be needed. It just calls `enrollment.api.add_enrollment()` which in turn calls `enrollment.data.create_course_enrollment` which in turn calls `CourseEnrollment.enroll(..., check_access=True)`. None of these methods check the embargo restrictions (presumably because they don't have the `request`). 
   Note: the `CourseEnrollment.enroll` method's comments say that it contains "All the server-side checks for whether a user is allowed to enroll." but that is not true, since it doesn't check embargo restrictions.
4. Now that the user is enrolled, the JavaScript redirects them to `/course_modes/choose/:course_id/`. This view *does* check for embargo restrictions, and presents an error message to the user.
5. However, if the user goes to the dashboard, they'll see that they're now enrolled in the course, although they won't be able to view the course because the courseware does check the embargo restrictions.


I think the easiest way to fix this will be to put an embargo check in the `baskets` API method, since that's currently the only server-side code in this flow that has the `request` (and thus the user's IP). Does that make sense?

rlucioni replied:
@bradenmacdonald thanks for looking into this. Adding an embargo check in the `baskets` API method (before checking for SKUs) sounds like the right approach. Since this issue currently exists in edx-platform, fixing it seems to be outside the scope of the work you're doing here. Let me know if you agree, and I'll make sure the bug gets tracked and prioritized internally.

bradenmacdonald replied:
@rlucioni Yes, could you please open a bug for that internally? I'm on a tight schedule for the SSO work and I agree that since it's an existing issue it's outside the scope of this PR. Thanks!

rlucioni commented on common/djangoapps/student/views.py:
```diff
@@ -357,24 +357,29 @@ def signin_user(request):
+    third_party_auth_error = None
+    for msg in messages.get_messages(request):
+        if msg.extra_tags.split()[0] == "social-auth":
+            third_party_auth_error = unicode(msg)
```
Will this allow untranslated error messages to leak into the system?

bradenmacdonald replied:
Potentially, yes. The template should always prefix the error with a translated message like "Sorry, that login didn't work:" and then this technical explanation which may not be translated. I guess I'm assuming that giving a specific reason, even if untranslated, along with a translated generic error will be more helpful than not.

rlucioni replied:
@frrrances and/or @srpearce, do you have any feedback on this? @bradenmacdonald, I'm wondering if that more technical explanation is something a user would want to see. It's helpful to developers, but as a user I might not know what to do about some of those messages.

bradenmacdonald replied:
@rlucioni I understand, but sometimes the user *does* want to see this message. For example, this is how the `NotActivatedException` (`_('This account has not yet been activated. An activation email has been re-sent.')`) gets passed on to the user. It could be frustrating to not know why the login didn't work without that explanation. In that case, it will be translated, since the exception message is defined in our code. Other messages may not be translated (since they're provided by external code), but they should hopefully be rare.

rlucioni replied:
Thanks for the context @bradenmacdonald.

rlucioni replied:
@bradenmacdonald, I agree with your approach to this. However, I want to make sure @sarina is aware that this has the potential to allow untranslated strings into the system, and that she's on board with it. @sarina?

sarina replied:
Yep, that's reasonable.

It may be safer `_(msg)`: that way if the message is in English and we somehow have that in our translation file, it'll get returned in the user's preferred language. Added benefit, `_()` will always return a unicode object.

bradenmacdonald replied:
@sarina: Cool, I'll do that :)

rlucioni commented:
@srpearce sure. The relevant lines are as follows:

1.[ Displayed when a user attempts to sign into an inactive edX account using third-party auth](https://github.com/edx/edx-platform/pull/8262/files#diff-5a31e91b78b2599dc5ace9c2a3b31d8cR183)
2. [Displayed when a third-party auth error occurs](https://github.com/edx/edx-platform/pull/8262/files#diff-b74ffc62cd4c2ba2c95084abd3ace1c7R164), along with what I believe is a relevant detail message derived from [python-social-auth's exceptions](https://github.com/omab/python-social-auth/blob/ee3c2c286e474c933fcd3d4bfe9df1d03204ad70/social/exceptions.py). @bradenmacdonald can you confirm this? 
3. [Displayed when the user signs in with a third-party auth account which isn't linked to an edX account](https://github.com/edx/edx-platform/pull/8262/files#diff-dc2d864ace23ab90abbe30c132eae553R4)

Let me know if you have any questions.

frrrances commented:
Thanks to @rlucioni I have a good sense of what's going on and I would recommend that you make the following changes to the loading screen: 

1. Provide some orientation/wayfinding for users to feel secure that they are still on the edX/Open edX site and haven't been redirected elsewhere. This can be as simple as an edX logo and using some of our site colors for the text (blue, gray). 
2. Remove the detailed list of things that are shown and just say loading or please wait. 
3. Provide some visual design. I can be available to review or give advice on Monday if you need help with this.

Lastly, I want to note a few things to consider about how this fits into the broader site and the user's experience: 
1. We should keep in mind that while things may be quick for people with fast connections, users with less bandwidth may spend some time here so we should not brush it of as mostly instantaneous.
2. We should try to be aware of how many places we ask users to "wait" while looking at a loading indicator. One is no big deal, but if we start to see one after another after another, it will add up to a bad experience. 

Thanks for working to make things more logical and let me know if I can help with advise on the design side of things. Let me know when you've had a chance to address those and I'll take another look for thumbs.

bradenmacdonald commented:
@frrrances Thanks. What do you mean by "Provide some visual design"?

nasthagiri commented:
@bradenmacdonald To verify that your changes will not break mobile, please execute the following curl statements against your open-edx instance:

- curl -X POST -d "client_id=<EDX_OAUTH_CLIENT_ID>&access_token=<THIRD_PARTY_OAUTH_TOKEN>" http://<SANDBOX>/oauth2/exchange_access_token/facebook/

  - This should return something like: 
`{"access_token": "<EDX_OAUTH_TOKEN>", "token_type": "Bearer", "expires_in": 2591999, "scope": ""}`

- curl -H "Authorization: Bearer <EDX_OAUTH_TOKEN>" http://<SANDBOX>/api/mobile/v0.5/users/<EDX_USERNAME>
  - This should return something like:
`{"id": 118, "username": "<EDX_USERNAME>", "email": "nasthagiri@edx.org", "name": "Nimisha Asthagiri", "course_enrollments": "http://<SANDBOX>/api/mobile/v0.5/users/<EDX_USERNAME>/course_enrollments/"}`


**Note:**
* If you haven't already, the `EDX_OAUTH_CLIENT_ID` is something you can create at the following location on your sandbox: `http://<SANDBOX>/admin/oauth2/client/`
* The `THIRD_PARTY_OAUTH_TOKEN` is something you can retrieve from the `social_auth_usersocialauth` SQL table once the handshake with the 3rd party auth provider is completed.

Let me know if you run into any issues or have any questions.  Thanks.

srpearce commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -192,6 +177,12 @@ class AuthEntryError(AuthException):
+class NotActivatedException(AuthException):
+    """ Raised when a user tries to login to an unverified account """
+    def __str__(self):
+        return _('This account has not yet been activated. An activation email has been re-sent.')
```
Is it possible to include the e-mail address where we're sending the message here? If so, I'd say "This account has not yet been activated. An activation email has been re-sent to `<address>`." If not, I'd just say "This account has not yet been activated. An activation email has been re-sent to the address associated with this account."

bradenmacdonald replied:
Sure:

![screen shot 2015-06-11 at 5 39 37 pm](https://cloud.githubusercontent.com/assets/945577/8121462/e8d56b1a-1060-11e5-9a04-d360fe24e038.png)

srpearce commented on lms/templates/login.html:
```diff
@@ -162,6 +159,15 @@ <h3 class="message-title">${_("We couldn't log you in.")} </h3>
 
+      % if third_party_auth_error:
+        <div role="alert" class="status message third-party-auth-error is-shown" tabindex="-1">
+          <h3 class="message-title">${_("We're sorry, that didn't work.")} </h3>
```
An error occurred signing you in to `<platform>`.

rlucioni replied:
@srpearce should this be "An error occurred *when* signing you in to `<platform>`."

bradenmacdonald replied:
Is this ok? The new text may be a bit long for the UI:

![screen shot 2015-06-11 at 5 27 52 pm](https://cloud.githubusercontent.com/assets/945577/8121332/388c96ee-105f-11e5-92ba-47397e176b5d.png)

srpearce commented:
@rlucioni I've provided some text suggestions. Let me know if you have any questions or comments.

bradenmacdonald commented:
@frrrances I made the UX changes you requested (at least points 1 and 2). Here's an updated screenshot (keep in mind the spinner is animated):

![new page screenshot](https://cloud.githubusercontent.com/assets/945577/8120534/082e5f78-1055-11e5-80fd-76793894ccbd.png)

I think this is nicer, and because the header is the same as the subsequent page (dashboard), it flows very smoothly and doesn't feel like a jarring redirect - instead it just feels like the "Please Wait" message gets replaced with the list of courses once it's loaded.

bradenmacdonald commented:
@nasthagiri Thanks for the instructions for testing the mobile app. I just went through that on my devstack (except with "google-oauth2" instead of "facebook" since I haven't set up Facebook), and it all worked as described.

bradenmacdonald commented:
@srpearce @rlucioni I updated the text strings: 336017533aa6cdf7b35a60a7e2997b85b1e3445d

frrrances commented:
@bradenmacdonald Thanks for those changes. From the screenshot, this looks much better. :+1:

rlucioni commented:
@frrrances, do you have a sense of whether the interstitial page being added is an accessibility concern?

@bradenmacdonald :+1: from me. Before this is merged, we still need agreement on an upgrade strategy from DevOps (@feanil) and an all clear on the open accessibility question. Thanks for your patience!

frrrances commented:
@rlucioni No, I would have @cptvitamin review for accessibility.

bradenmacdonald commented:
@rlucioni I was just about to ping you and ask for a final engineering +1 but you beat me to it - thank you very much! I did *just* push a commit though: b12e24c8f1e1c7643b051ae2fbda5096b462ed63 - it adds the timeout that was just discussed on the [CRI-9](https://openedx.atlassian.net/browse/CRI-9) JIRA ticket. I have tested it manually by configuring the timeout to be very short and it seems to work well. I'm not sure if you saw that commit yet or not, but if not can you please take a look? I think that should be it.

rlucioni commented:
@bradenmacdonald I see it now, and I read through the thread in [CRI-9](https://openedx.atlassian.net/browse/CRI-9). Looks good to me. :+1: 

Thanks for your hard work on this!

srpearce commented on lms/templates/login.html:
```diff
@@ -162,6 +159,15 @@ <h3 class="message-title">${_("We couldn't log you in.")} </h3>
 
+      % if third_party_auth_error:
+        <div role="alert" class="status message third-party-auth-error is-shown" tabindex="-1">
+          <h3 class="message-title">${_("An error occurred when signing you in to {platform_name}.").format(platform_name=platform_name)} </h3>
```
It's grammatically correct with or without the "when". Personally, I think it sounds better without.

srpearce commented on lms/templates/student_account/login.underscore:
```diff
@@ -1,7 +1,7 @@
     <% if (context.currentProvider) { %>
         <p class="message-copy">
-            <%- _.sprintf( gettext("You've successfully signed into %(currentProvider)s, but your %(currentProvider)s account isn't linked with an %(platformName)s account. To link your accounts, go to your %(platformName)s account settings."), context ) %>
+            <%- _.sprintf( gettext("You have successfully signed into %(currentProvider)s, but your %(currentProvider)s account is not linked with any %(platformName)s account. To link your accounts, sign in now using your %(platformName)s password."), context ) %>
```
I used "this" because, although the user hasn't signed in to an edX account, he _is_ trying to sign in to a specific account. "Any" sounds a little harsh to me. Since the main thing we're trying to avoid is having to switch between a/an in front of "your platform name", what about this?

You have successfully signed into %(currentProvider)s, but your %(currentProvider)s account does not have a linked %(platformName)s account. To link your accounts, sign in now using your %(platformName)s password.

bradenmacdonald replied:
Sounds good to me.

nasthagiri commented on common/djangoapps/third_party_auth/pipeline.py:
```diff
@@ -445,13 +427,46 @@ def parse_query_params(strategy, response, *args, **kwargs):
+def reset_pipeline_timeout(strategy, *args, **kwargs):
+    """ Revert the short session timeout now that the pipeline is completed. """
+    if strategy.request and strategy.request.session.pop('third_party_pipeline_timeout_active', False):
+        strategy.request.session.set_expiry(None)  # None makes the session revert to the global session expiry policy.
```
hmm.. please note that [elsewhere](https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/views.py#L1087) in the codebase we `set_expiry` either to `0` (session expires when the browser session is closed) or to `604800` (if remember-me was selected). 

Will this change override that behavior?

bradenmacdonald replied:
Thanks @nasthagiri, good catch. It turns out that we don't need this `reset_pipeline_timeout` step at all - because we only set a shortened timeout if no user is logged in and no edx user is linked to the third party account. In that case, the user must necessarily either do a normal login or register a new account. Either of those actions will set a new expiry time, as you've just pointed out.

I've pushed a fix and tested it.

sarina commented:
@bradenmacdonald @rlucioni what's the status of this pull request?

rlucioni commented:
@sarina we still need an accessibility review and agreement on an upgrade strategy from DevOps.

feanil commented:
@rlucioni I believe we have already talked about doing the fake migration when it comes time to deploy this code to edx production environments.  Is there something more, that I'm not aware of?

rlucioni commented:
Great, thanks @feanil. I was making sure we were all set on that front.

@bradenmacdonald and @sarina, all we need before merging this is an accessibility review from @cptvitamin. I'll do my best to help him walk through the changes being made today.

cptvitamin commented:
@bradenmacdonald, @rlucioni just walked me through this process and it appears as though the only event that would have an accessibility impact would be the interstitial page that asks a user to "Please wait"  Do you agree with that?  If so, I think this should be fine.  The page's ```<title>``` element, the first thing read to a screen reader user, indicates as much, and if the page was onscreen long enough for the screen reader user to explore it, they first Heading says the same thing.  I do not think anything like a ```role=alert``` is necessary here and would just complicate things.  That is my long winded way of saying :+1: 

Unless there is some other UI you think I should take a look at.

rlucioni commented:
@bradenmacdonald I'm ready to merge this once you respond to @cptvitamin's comment.

bradenmacdonald commented:
@cptvitamin Yep, I agree that that's the only accessibility impact. Thanks!

@rlucioni Please merge away :) Thanks a lot for all your help on this. Do you want [me] to squash this first though?

sarina commented:
@rlucioni I confirmed with Nimisha that Mobile is good with this PR. We've got UX and a11y sign off as well. Can you respond to Braden above, and merge when it's looking good to you? Thanks!

rlucioni commented:
Thanks for the follow-up @sarina. @bradenmacdonald yes, please squash your commits. As I said earlier, since your changes require an upgraded version of python-social-auth (which may not be backward-compatible with the old code), I think the version change should be bundled together with your other work.

bradenmacdonald commented:
@rlucioni I fixed the JS issues: The fixes I made are in cbdb8526161b6cf5edcdca5a0403409b3014308d but I subsequently squashed them into the main commit. Please take a quick look at that commit, and then this should be ready to merge.

rlucioni commented on lms/static/js/student_account/views/AccessView.js:
```diff
@@ -1,19 +1,11 @@
     edx.student.account = edx.student.account || {};
 
-    // Bind to StateChange Event
-    History.Adapter.bind( window, 'statechange', function() {
```
@bradenmacdonald why is this being removed? It makes the back button compatible with the logistration page.

bradenmacdonald replied:
As discussed via chat: JSHint was complaining about this code and it appears to not be doing anything. Let me know if you find out otherwise, and I can revert this since it's mostly an incidental change.

rlucioni replied:
I did a little more digging into the event handler, and I agree that it's not necessary. [`getState()`](https://github.com/browserstate/history.js/#states) doesn't appear to have any side effects.

sarina commented:
Thanks Renzo!!!

Braden - could you share with me any SSO project plan? (fine to do in
email). I want to understand where the SSO project is at and what other PRs
may be coming in for it.

On Fri, Jun 19, 2015 at 9:12 AM, Renzo Lucioni <notifications@github.com>
wrote:

> Merged #8262 <https://github.com/edx/edx-platform/pull/8262>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8262#event-335222760>.
>

rlucioni commented on common/djangoapps/student/views.py:
```diff
@@ -874,20 +849,6 @@ def change_enrollment(request, check_access=True):
         # Otherwise, there is only one mode available (the default)
         return HttpResponse()
-
-    elif action == "add_to_cart":
```
@bradenmacdonald could you explain why this was removed? FYI @chrisndodge.

rlucioni replied:
@bradenmacdonald I see that the functionality moved to the auth completion view, which makes sense given your refactor. The question now is why the shoppingcart addition isn't working for @chrisndodge, even though the new code is tested.

bradenmacdonald replied:
@rlucioni Yes, since the new auth completion view uses the [enrollment API](https://github.com/edx/edx-platform/tree/master/common/djangoapps/enrollment), since no other edx-platform code seemed to use this functionality here now that I did that refactor, and since the enrollment API is available for external enrollment calls via API, it seems there is no need to leave this code here.

Can you give me a link to an example on staging or instructions for reproducing the issue?

rlucioni replied:
@bradenmacdonald I'm not familiar with the reproduction steps for this issue, but @chrisndodge should have more details.

bradenmacdonald replied:
@rlucioni EDIT: What I said before didn't really make sense. But yeah, the functionality is available in the auth completion view instead. It says right in the comment that "this elif clause is only used in the "auto-add after user reg/login" case, i.e. it's always wrapped in try_change_enrollment.", and I removed `try_change_enrollment`.

Can you give me a link to an example on staging or instructions for reproducing the issue?

chrisndodge replied:
Just closing the loop on the added discussion on this PR, post merge. The issue is that there was a change to the JavaScript on the registration page. Since the JavaScript is in the .html template file (rather than a separate .js file), the JS changes have to be also applied to our customers customized registration (it is customized since they have additional signup fields to capture). This has been applied and verified on staging. 

As follow-up work, it'd be nice to get the JavaScript out of the the registration.html template and into it's own file at some point.

rlucioni commented on lms/static/js/student_account/views/FinishAuthView.js:
```diff
@@ -0,0 +1,170 @@
+                    /* Attempt to auto-enroll the user in a free mode of the course,
+                    then redirect to the next location. */
+                    enrollmentInterface.enroll( courseId, redirectUrl );
+                } else if ( this.enrollmentAction === 'add_to_cart' && this.courseId) {
```
Looks like the functionality moved here.
